docs(uk): translate hooks.md to Ukrainian (#598)
reactjs/uk.react.dev: src/content/reference/react/hooks.md

<Intro>

> *Хуки* дають змогу використовувати різноманітні можливості React усередині компонентів. Ви можете застосовувати як вбудовані хуки, так і поєднувати їх для створення власних. На цій сторінці перелічено всі хуки, що вбудовані в React.

Author: Тут не згоден. Як розробники ми використовуємо можливості Реакта, а не його функції.

> *Хуки* дають змогу використовувати різноманітні можливості React усередині компонентів. Ви можете застосовувати як вбудовані хуки, так і поєднувати їх для створення власних. На цій сторінці перелічено всі хуки, що вбудовані в React.

Author: ок.
Тоді не функції. а "функціонал". Щоб відокремити функції як такі. 
Хуки дозволяють використовувати різний функціонал React у ваших компонентах.

> *Хуки* дають змогу використовувати різноманітні можливості React усередині компонентів. Ви можете застосовувати як вбудовані хуки, так і поєднувати їх для створення власних. На цій сторінці перелічено всі хуки, що вбудовані в React.

Reviewer: щодо функціонала є гарний сайт https://devium.me/functional/ :)
функціональність - ок, з цим можу погодитися

> *Хуки* дають змогу використовувати різноманітні можливості React усередині компонентів. Ви можете застосовувати як вбудовані хуки, так і поєднувати їх для створення власних. На цій сторінці перелічено всі хуки, що вбудовані в React.

Author: ок) варіанти
Хуки дозволяють використовувати різну функціональність React у ваших компонентах.
Хуки дають змогу компонуванню функціональності React у ваших компонентах.
Хуки відкривають доступ до різної функціональності React зсередини ваших компонентів.

> *Хуки* дають змогу використовувати різноманітні можливості React усередині компонентів. Ви можете застосовувати як вбудовані хуки, так і поєднувати їх для створення власних. На цій сторінці перелічено всі хуки, що вбудовані в React.

Reviewer: прямий переклад і все

```suggestion
*Хуки* дають змогу використовувати різноманітну функціональність React усередині компонентів. Ви можете застосовувати як вбудовані хуки, так і поєднувати їх для створення власних. На цій сторінці перелічено всі хуки, що вбудовані в React.
```

</Intro>

---

## Хуки стану {/*state-hooks*/}

*Стан* дає компоненту змогу ["запам’ятовувати" інформацію таку як, наприклад, введення користувача.](/learn/state-a-components-memory) Скажімо, компонент форми може зберігати введене значення, а компонент галереї зображень — індекс вибраного зображення.

Щоб додати стан до компонента, використовуйте один із цих хуків:

* [`useState`](/reference/react/useState) оголошує змінну стану, яку ви можете оновлювати безпосередньо.
* [`useReducer`](/reference/react/useReducer) оголошує змінну стану з логікою оновлення всередині [функції-редюсера.](/learn/extracting-state-logic-into-a-reducer)

```js

> function ImageGallery() {

Reviewer: Переглянь, будь ласка, ще раз файл, щоб було рядок у рядок, а то щось мої виправлення не допомогли

Author: Як зробити це технічно правильно? Щоб вирівняти рядок-у-рядок...

Reviewer: я не знаю, де перекладаєш - я відкриваю у vscode зміни і дивлюся, і там видно, наприклад:
![Screenshot from 2025-04-06 19-32-46](https://github.com/user-attachments/assets/d7a6f459-1830-4431-bd46-87966f83710f)

Author: Це я розумію.
Тобто мені вирівняти все рядок-в-рядок і ще раз зробити pull?
В цьому pull виправити тільки це?

Author: Оці всі зіміни вище виправляти не треба?

---

## Хуки рефів {/*ref-hooks*/}

> *Рефи* дають компоненту змогу [зберігати довільну інформацію, яка не бере участі в рендері.](/learn/referencing-values-with-refs) Наприклад, DOM-вузол або ідентифікатор таймауту. На відміну від стану, оновлення рефа не призводить до повторного рендеру компонента. Рефи — це "рятівне рішення" за межами парадигми React. Вони корисні, коли потрібно працювати з не-React системами, як-от вбудовані API браузера.

Author: Тут я навмисно розбив довге реченя на два. На англійській мові "Refs let a component hold some information that isn't used for rendering, like a DOM node or a timeout ID" занадто _навантажене_, як на мене.  А переклад на українську ще важчий - занадто багато ком. Пропоную лишити як є - перше речення що таке речення. Друге - "наприклад"... Але не проти вашого варіанту.

> *Рефи* дають компоненту змогу [зберігати довільну інформацію, яка не бере участі в рендері.](/learn/referencing-values-with-refs) Наприклад, DOM-вузол або ідентифікатор таймауту. На відміну від стану, оновлення рефа не призводить до повторного рендеру компонента. Рефи — це "рятівне рішення" за межами парадигми React. Вони корисні, коли потрібно працювати з не-React системами, як-от вбудовані API браузера.

Reviewer: ні, тут немає варіантів, перекладаємо речення у речення

* [`useRef`](/reference/react/useRef) оголошує реф. Ви можете зберігати в ньому будь-які дані, найчастіше — посилання на DOM-вузол.

> * [`useImperativeHandle`](/reference/react/useImperativeHandle) дає змогу кастомізувати реф, який ваш компонент передає назовні. Використовується рідко.

Author: Я читаю документацію Реакт і, знаєте, вона, як то правильно сказати... вона не академічна, не енциклопедична. Це не документація "для чайників", але для тих. хто шарить щось в програмуванні і інтерфейсах. Вона, документація, легка і невимушена. "Налаштувати власний" в цьому сенсі програє перед "кастомізувати".

> * [`useImperativeHandle`](/reference/react/useImperativeHandle) дає змогу кастомізувати реф, який ваш компонент передає назовні. Використовується рідко.

Reviewer: проблема не академічний чи ні, а засилля англіцизмів чи ні. намагаємося якнайменше використовувати англіцизми за можливості

> * [`useImperativeHandle`](/reference/react/useImperativeHandle) дає змогу кастомізувати реф, який ваш компонент передає назовні. Використовується рідко.

Author: прийнято.
але "власний" опускаємо
... дає змогу налаштувати реф ...

> * [`useImperativeHandle`](/reference/react/useImperativeHandle) дає змогу кастомізувати реф, який ваш компонент передає назовні. Використовується рідко.

Reviewer: ```suggestion
* [`useImperativeHandle`](/reference/react/useImperativeHandle) дає змогу налаштувати реф, який ваш компонент передає назовні. Використовується рідко.
```

  // ...
```

> Ефекти — це "аварійний вихід" за межі парадигми React. Не використовуйте ефекти для керування потоком даних у вашому застосунку. Якщо ви не взаємодієте із зовнішньою системою, [можливо, ефект вам взагалі не потрібен.](/learn/you-might-not-need-an-effect)

Author: Мммм... то ж якщо вам сподобалось (мені теж) "рятівне рішення", давайте й тут ось так, наприклад, сформулюємо "Ефекти — це рятівне рішення поза межами парадигми React."

> Ефекти — це "аварійний вихід" за межі парадигми React. Не використовуйте ефекти для керування потоком даних у вашому застосунку. Якщо ви не взаємодієте із зовнішньою системою, [можливо, ефект вам взагалі не потрібен.](/learn/you-might-not-need-an-effect)

Reviewer: ```suggestion
Ефекти — це "рятівне рішення" за межі парадигми React. Не використовуйте ефекти для керування потоком даних у вашому застосунку. Якщо ви не взаємодієте із зовнішньою системою, [можливо, ефект вам взагалі не потрібен.](/learn/you-might-not-need-an-effect)
```

> Існують дві рідковживані варіації `useEffect`, які відрізняються моментом виконання:

Author: Ось тут просто "вау"! Звичайно "є" - просто, легше, природньо!

> * [`useLayoutEffect`](/reference/react/useLayoutEffect) спрацьовує до того, як браузер перемалює екран. У цей момент можна дізнатися розміри або положення елементів у DOM.

Reviewer: ```suggestion
* [`useLayoutEffect`](/reference/react/useLayoutEffect) спрацьовує до того, як браузер перемалює екран. У ньому можна виміряти компонування елементів.
```

Author: )) Памятаю як з GPT перекладали цей фрагмент.
В англійському варіанті "You can measure layout here." класно ловить контекст першого речення і класно розкриває суть хука. Ось це "measure layout" - тут дуже влучно, коротко і зрозуміло. І от український варіант "У ньому можна виміряти компонування елементів" - близько до оригіналу, але, як на мене, занадто абстрактно.
Тому я вирішив в цю абстракцію внести трошки конкретики.
Ще ось такі варіанти:
У ньому можна виміряти розміри й положення елементів у DOM.
У цей момент можна виміряти їхнє компонування - розміри й положення у DOM.
Хоч в англійському варіанті безпосередньо ДОМ не згадується, але ось цей "measure layout" говорить саме про вимірювання розмірів і положення елементів у ДОМ.
Але я не проти вашого варіанту. Лише навожу свої аргументи і ще раз підкреслюю, що він занадто абстрактний. Як на мене ;)

Reviewer: це документація, її треба перекладати близько до оригіналу за можливості, змінюючи лише те, що не притаманно стилю мови. наприклад, замість притаманних англійській повторень в українській можна викидати і замінювати на тире. або ж тут here (у ньому, тут) це зовсім не те, що `у цей момент`, бо це вже зовсім інше пояснення - тут важливо зберігати такі, на перший погля не важливі, технічні речі

якщо не зрозуміло або слово нове вводиться, то можна дійсно розширити визначення (наприклад, не просто компонування, а компонування елементів) або/і додати оригінал чи пояснення в дужках. суть у тому, щоб можна було реконструювати оригінал або звернутися до нього за необхідності. прибираючи компонування як таке, ми прибираємо термін. прибираючи термін, ми повертаємося до того, що у майбутньому людям не зрозуміло, що таке компонування і треба щоразу повноцінно пояснювати

```suggestion
* [`useLayoutEffect`](/reference/react/useLayoutEffect) спрацьовує до того, як браузер перемалює екран. У ньому можна виміряти компонування елементів (layout).
```

* [`useInsertionEffect`](/reference/react/useInsertionEffect) спрацьовує до того, як React змінить DOM. Бібліотеки можуть вставляти динамічний CSS саме тут.

---

## Хуки продуктивності {/*performance-hooks*/}

> Поширений спосіб оптимізації повторного рендеру — уникати зайвих обчислень. Наприклад, ви можете вказати React повторно використати кешований результат або пропустити ререндер, якщо дані не змінилися з моменту попереднього рендеру.

Reviewer: ```suggestion
Поширений спосіб оптимізації повторного рендеру — уникати зайвих обчислень. Наприклад, ви можете вказати React повторно використати кешований результат або пропустити повторний рендер, якщо дані не змінилися після попереднього.
```

Author: Я об'єднав би свій варіант з вашим зауваженням.
"Ререндер" - сталий вираз серед фронтендерів "повторний рендер" - як на мене, дуже академічно і сухо.
А ваша кінцівка речення "якщо дані не змінилися після попереднього" - дуже класно.

Reviewer: тут знову проблема англіцизмів, а не академічності, а ще межі прескриптивізму, беручи до уваги, що рендер - це вже запозичення.

re-render як повторний рендер залишилося від попередньої команди - і я поки з цим згодна. ре - це англійський префікс, не український, навіть за бажання скорити, то краще вже перерендер. уявіть, що документацію читають люди, які не знають англійську взагалі - похідні слова від рендеру мають бути зрозумілі цій українськомовній людині

> Щоб уникати обчислень і непотрібного рендерингу, скористайтеся одним із цих хуків:

Reviewer: ```suggestion
Щоб уникати зайвих обчислень чи повторного рендерингу, скористайтеся одним із цих хуків:
```

Author: В англійському варіанті "To skip calculations and **unnecessary** **re-rendering**, use one of these Hooks:"
Дозволю собі відмітити, що у вашому варіанті дещо втрачений сенс, що є в оригіналі.
Хуки про які йдеться нижче вони уникають обчислень, а не "повторного рендерингу". Повторний рендерінг - це як наслідок обчислень, а не альтернатива. Тобто "чи повторного..." тут не підходить.
В англійському варіанті слово "unnecessary" стосується саме ререндерингу, а не обчислень. Ви тут змістили акцент.
Можливо вам здається мій варіант не зовсім вдалим.
Ось ще варіанти
Щоб уникати зайвих обчислень і непотрібного повторного рендерингу...
Щоб уникати зайвих обчислень, що призводять до повторного рендерингу...

Reviewer: ```suggestion
Щоб уникати зайвих обчислень і непотрібного повторного рендерингу, скористайтеся одним із цих хуків:
```

- [`useMemo`](/reference/react/useMemo) дає змогу кешувати результат ресурсомісткого обчислення.

> - [`useCallback`](/reference/react/useCallback) дає змогу кешувати функцію перед її передаванням до оптимізованого компонента.

Reviewer: ```suggestion
- [`useCallback`](/reference/react/useCallback) дає змогу кешувати визначення функції перед її передаванням до оптимізованого компонента.
```

Author: Так)) Тут теж доводилось залізати "під капот" і розбиратись що саме кешує useCallback - функцію, визначення, результат виконання... Хоч англійське "function definition"  - це "визначення функції", але це тільки якщо буквально. .Бо в контексті Реакт воно скоріше означає саму функцію як об'єкт (з урахуванням замикань, контексту, залежностей).
Тобто, ось в реакт компоненті є функція, яку ми обгорнули в useCallback. React на першому рендері створює функцію в себе й кешує її посилання. Якщо залежності не змінилися, посилання зберігається між рендерами, і нова функція не створюється. Тобто **кешується саме функція як значення**, а не її текстове визначення, не ось це буквальне "definition".
Тому "визначення функції" звучить технічніше, але втрачається широта того всього, що кешується, ніж коли використати просто "функцію".

Reviewer: якщо є проблема в оригіналі, то додавай issue їм в оригінал.
тут переклади визначення функції, але можеш додати в дужках примітку перекладача про цей технічний аспект

> 

Reviewer: ```suggestion
```

}
```

> Інколи уникнути ререндеру неможливо, бо екран справді потребує оновлення. У такому випадку продуктивність можна покращити, розділивши блокувальні оновлення, які мають бути синхронними (наприклад, введення тексту в поле), і неблокувальні, які не обов’язково затримують інтерфейс (наприклад, оновлення діаграми).

Reviewer: ```suggestion
Інколи уникнути повторного рендерингу неможливо, бо екран справді потребує оновлення. У цьому разі продуктивність можна покращити, розділивши блокувальні оновлення, які мають бути синхронними (наприклад, введення тексту в поле), і неблокувальні, які не обов'язково затримують інтерфейс (наприклад, оновлення діаграми).
```
- https://onlinecorrector.com.ua/%D1%83-%D1%82%D0%B0%D0%BA%D0%BE%D0%BC%D1%83-%D1%86%D1%8C%D0%BE%D0%BC%D1%83-%D1%82%D0%BE%D0%BC%D1%83-%D1%80%D0%B0%D0%B7%D1%96/
- використовуємо цей апостроф `'`

Author: Зі всім згоден окрім "цьому" ))).
У цьому - це якщо мова йде про конкретний випадок, коли ми щось обговорюємо контректне. А якщо ми говоримо про щось подібне, вважаю краще "такий". Тобто в в цьому випадку пропоную використати "у такому разі"

Reviewer: ```suggestion
Інколи уникнути повторного рендерингу неможливо, бо екран справді потребує оновлення. У такому разі продуктивність можна покращити, розділивши блокувальні оновлення, які мають бути синхронними (наприклад, введення тексту в поле), і неблокувальні, які не обов'язково затримують інтерфейс (наприклад, оновлення діаграми).
```

Щоб керувати пріоритетом рендерингу, скористайтеся одним із цих хуків:

- [`useTransition`](/reference/react/useTransition) дає змогу позначити зміну стану як неблокувальну, дозволяючи іншим оновленням її переривати.

> - [`useDeferredValue`](/reference/react/useDeferredValue) дає змогу відкласти оновлення неважливої частини інтерфейсу, щоб інші частини могли оновитися раніше.

Reviewer: ```suggestion
- [`useDeferredValue`](/reference/react/useDeferredValue) дає змогу відкласти оновлення неважливої частини інтерфейсу, щоб спершу могли оновитися інші частини.
```

---

## Інші хуки {/*other-hooks*/}

> Ці хуки здебільшого корисні авторам бібліотек й рідко використовуються у звичайному коді застосунків.

Reviewer: ```suggestion
Ці хуки здебільшого корисні авторам бібліотек і рідко використовуються у звичайному коді застосунків.
```

Author: Тут справа смаку. З "й" речення більш акцентоване на "рідко". З "і" - чотири букви "і"на 15 символів. Вирішуйте самі. Мені з "й" більше подобається.

Reviewer: це не бажання, це правила чергування https://ukr-mova.in.ua/library/orfografiya/cherguvannya-i-y-ta
між приголосними ставимо і

> - [`useDebugValue`](/reference/react/useDebugValue) дає змогу кастомізувати мітку, яку React DevTools показує для вашого кастомного хука.

Reviewer: ```suggestion
- [`useDebugValue`](/reference/react/useDebugValue) дає змогу налаштувати мітку, яку React DevTools показує для вашого власного хука.
```

Reviewer: знову англіцизми

я за хуки користувача, але тут проблема більше у фразуванні "ваш хук користувача", коли цим користувачем є ви, тому вибрала "ваш власний хук". я знаю за кастомність і похідні, сама вживаю, для цього перекладу не використовуємо. 

чому взагалі власний для custom у мене в пріоритеті, бо власний - це сам написав, сам володію, за нього відповідаю, мені властивий. і різниці не дуже багато між your own hook і your custom hook. ну і коротке. серед варіантів ще є: саморобний, створений/налаштований власноруч/самотужки/самостійно

Author: так, я вже зрозумів стосовно "кастомних хуків" - використовуємо "хуки користувача"

Reviewer: і тут також най буде

```suggestion
- [`useDebugValue`](/reference/react/useDebugValue) дає змогу налаштувати мітку, яку React DevTools показує для вашого хука користувача.
```

а потім в тому issue https://github.com/reactjs/uk.react.dev/pull/598#discussion_r2030209627 розберемося

> - [`useId`](/reference/react/useId) дає компоненту змогу прив’язати до себе унікальний ідентифікатор. Зазвичай використовується з API доступності.

Reviewer: ```suggestion
- [`useId`](/reference/react/useId) дає компоненту змогу прив'язати до себе унікальний ідентифікатор. Зазвичай використовується з API доступності.
```

> - [`useSyncExternalStore`](/reference/react/useSyncExternalStore) дозволяє компоненту підписатися на зовнішнє сховище.

Reviewer: ```suggestion
- [`useSyncExternalStore`](/reference/react/useSyncExternalStore) дає компоненту змогу підписатися на зовнішнє сховище.
```

- [`useActionState`](/reference/react/useActionState) дає змогу керувати станом дій.

---

## Власні хуки {/*your-own-hooks*/}

> Ви також можете [створювати власні кастомні хуки](/learn/reusing-logic-with-custom-hooks#extracting-your-own-custom-hook-from-a-component) як звичайні функції JavaScript.

Reviewer: ```suggestion
Ви також можете [створювати власні хуки користувача](/learn/reusing-logic-with-custom-hooks#extracting-your-own-custom-hook-from-a-component) як звичайні функції JavaScript.
```

Author: Ви також можете створювати **власні хуки користувача** - як на мене важко і занадто офіційно як для документації Реакт.
Я тут більше звернув увагу на неточність власного перекладу "власні кастомні" - тавтологія бо власний=кастомний. Тут в англійській мові "your own custom Hooks" слово  own підкреслює не власність, а те, що ви теж можете створювати такі хуки.
Тому і в вашій пропозиції "створювати власні хуки користувача" можна скоротити до "створювати хуки користувача". Бо не буває такого, що ви можете створювати "сторонні" (чи ще якісь) хуки окрім власних. Тому слово "власні" тут зайве.
Пропозиція "Ви також можете створювати кастомні хуки як звичайні функції JavaScript" - легко, просто і зрозуміло.

Reviewer: я пропоную зупинитися на https://github.com/reactjs/uk.react.dev/pull/598#discussion_r2029941420, а з цим питанням відкрити issue і потім вирішити

Author: згідно правил транслітерації - однозначно має бути гук))

Reviewer: # 606
